### PR TITLE
Create extension host processes from a node worker in the main process

### LINKED ISSUE #123592: Move extension host out of the workbench for process reuse
As part of #120431, we have to move the extension host out of the workbench.
Currently, the workbench is the parent of the extension host. When we do a reload, the workbench is recreated. When process reuse is enabled, the workbench PID stays the same throughout the reload, but the node environment is still taken down. Due to this cleanup, the child exit signal from the extension host is not handled, creating a zombie.
By moving the extension host out of the workbench, another longer-lived process can handle the signal instead.

@alexdima do you have any specific ideas as to how the new architecture should look?

### PR DESCRIPTION
**Pull request by alexdima**
Fixes #123592: Creates a node worker in the main process which creates extension host processes.

### CODE REVIEW

**Comment by bpasero** [comment]
@alexdima  looks a lot better now in CI, 👏 . maybe as a next step you restore the worker approach for forking the extension host?

File: src/vs/workbench/services/extensions/electron-sandbox/extensionHostStarter.ts [hunk 1]
```diff
@@ -3,7 +3,13 @@
  *  Licensed under the MIT License. See License.txt in the project root for license information.
  *--------------------------------------------------------------------------------------------*/
 
-import { registerSharedProcessRemoteService } from 'vs/platform/ipc/electron-sandbox/services';
+import { registerMainProcessRemoteService, registerSharedProcessRemoteService } from 'vs/platform/ipc/electron-sandbox/services';
 import { IExtensionHostStarter, ipcExtensionHostStarterChannelName } from 'vs/platform/extensions/common/extensionHostStarter';
 
-registerSharedProcessRemoteService(IExtensionHostStarter, ipcExtensionHostStarterChannelName, { supportsDelayedInstantiation: true });
+const location = 'main' as 'main' | 'shared';
```
[the remark this answers is not in the document]
Comment (alexdima) [reply src/vs/workbench/services/extensions/electron-sandbox/extensionHostStarter.ts:9]: I wanted to have an easy way to switch between the main process spawning the extension host and the shared process spawning the extension host. Currently, this means the extension host is spawned from the main process.

File: src/vs/platform/extensions/electron-main/workerMainProcessExtensionHostStarter.ts [hunk 2]
```diff
@@ -0,0 +1,160 @@
...
+		// Intentionally not killing the extension host processes
+	}
+
+	async _initialize(): Promise<void> {
+		this._proxy = await this._worker.getProxyObject();
+		this._logService.info(`ExtensionHostStarterWorker created`);
+	}
+
+	onDynamicStdout(id: string): Event<string> {
+		return this._proxy!.onDynamicStderr(id);
```
[the remark this answers is not in the document]
Comment (alexdima) [reply src/vs/platform/extensions/electron-main/workerMainProcessExtensionHostStarter.ts:110]: Good catch!

File: src/vs/platform/extensions/electron-main/directMainProcessExtensionHostStarter.ts [hunk 3]
```diff
@@ -0,0 +1,29 @@
+/*---------------------------------------------------------------------------------------------
+ *  Copyright (c) Microsoft Corporation. All rights reserved.
+ *  Licensed under the MIT License. See License.txt in the project root for license information.
+ *--------------------------------------------------------------------------------------------*/
+
+import { ExtensionHostStarter, IPartialLogService } from 'vs/platform/extensions/node/extensionHostStarter';
+import { ILifecycleMainService } from 'vs/platform/lifecycle/electron-main/lifecycleMainService';
+import { ILogService } from 'vs/platform/log/common/log';
+
+export class DirectMainProcessExtensionHostStarter extends ExtensionHostStarter {
```
Comment (bpasero) [inline src/vs/platform/extensions/electron-main/directMainProcessExtensionHostStarter.ts:10]: This file does not seem to be used anymore but I understand that you probably want to keep it for testing the starter without worker.

File: src/vs/platform/extensions/electron-main/workerMainProcessExtensionHostStarter.ts [hunk 4]
```diff
@@ -0,0 +1,160 @@
...
+			if (this._proxy) {
+				this._proxy.killAllNow();
+			}
+		});
+
+		// Normal shutdown: gracefully await extension host shutdowns
+		lifecycleMainService.onWillShutdown((e) => {
+			this._shutdown = true;
+			if (this._proxy) {
+				e.join(this._proxy.waitForAllExit(6000));
```
Comment (bpasero) [inline src/vs/platform/extensions/electron-main/workerMainProcessExtensionHostStarter.ts:95]: Just a nit, but we seem to be giving extensions `5s` for graceful shutdown from within the extension host but we seem to be waiting `6s` from the outside.

File: src/vs/platform/extensions/node/extensionHostStarter.ts [hunk 5]
```diff
@@ -115,17 +122,32 @@ class ExtensionHostProcess extends Disposable {
...
 		this._process.kill();
 	}
+
+	async waitForExit(maxWaitTimeMs: number): Promise<void> {
+		if (!this._process) {
+			return;
+		}
+		const pid = this._process.pid;
+		this._logService.info(`Waiting for extension host with pid ${pid} to exit.`);
+		await Promise.race([Event.toPromise(this.onExit), timeout(maxWaitTimeMs)]);
```
Comment (bpasero) [inline src/vs/platform/extensions/node/extensionHostStarter.ts:132]: To clarify: we are not explicitly terminating the extension host here because we assume that the window is closing which will close the socket and then should trigger the self-termination from within the extension host?

I wonder if maybe a more explicit flow should be implemented where you terminate the extension host from here, because once we move to a different IPC mechanism in the future, maybe we cannot rely on the socket-close event anymore.

File: src/vs/platform/extensions/electron-main/directMainProcessExtensionHostStarter.ts (hunk 3 above)
Comment (alexdima) [reply src/vs/platform/extensions/electron-main/directMainProcessExtensionHostStarter.ts:10]: :+1: Yes, there is also some code in the shared process that sets up an `IExtensionHostStarter` service that can be removed once we get confidence that spawning from main is the best way forward.

Comment (alexdima) [reply src/vs/platform/extensions/electron-main/directMainProcessExtensionHostStarter.ts:10]: I've created https://github.com/microsoft/vscode/issues/137025 to remember to do it.

File: src/vs/platform/extensions/electron-main/workerMainProcessExtensionHostStarter.ts (hunk 4 above)
Comment (alexdima) [reply src/vs/platform/extensions/electron-main/workerMainProcessExtensionHostStarter.ts:95]: Yes, I think this time needs to be >5s. If the main process receives a shutdown request via `onWillShutdown` and it begins killing renderer processes, IMHO it is possible that the extension host might still take some time >0 until it realizes that the renderer socket is closed and then begins the shutdown. We can maybe shorten this to 5.5s, but I would like to keep this >5s.

File: src/vs/platform/extensions/node/extensionHostStarter.ts (hunk 5 above)
Comment (alexdima) [reply src/vs/platform/extensions/node/extensionHostStarter.ts:132]: Yes, we need to revisit the shutdown of the extension host. Today, the extension host shuts down when the renderer sends it a shutdown message via the socket or when the socket closes. If the new IPC mechanism does not support closing we need to find an alternative way to shutdown the extension host (maybe by posting a message using nodejs to it).

File: src/vs/platform/extensions/electron-main/workerMainProcessExtensionHostStarter.ts (hunk 4 above)
Comment (bpasero) [reply src/vs/platform/extensions/electron-main/workerMainProcessExtensionHostStarter.ts:95]: Thanks for the explanation makes sense.
